src/datasets/features/audio.py

```diff
@@ -282,19 +287,44 @@ def _decode_non_mp3_file_like(self, file, format=None):
     def _decode_mp3(self, path_or_file):
         try:
             import torchaudio
-            import torchaudio.transforms as T
         except ImportError as err:
-            raise ImportError(
-                "Decoding 'mp3' audio files, requires 'torchaudio<0.12.0': pip install 'torchaudio<0.12.0'"
-            ) from err
-        if not version.parse(torchaudio.__version__) < version.parse("0.12.0"):
-            raise RuntimeError(
-                "Decoding 'mp3' audio files, requires 'torchaudio<0.12.0': pip install 'torchaudio<0.12.0'"
-            )
-        try:
-            torchaudio.set_audio_backend("sox_io")
-        except RuntimeError as err:
-            raise ImportError("To support decoding 'mp3' audio files, please install 'sox'.") from err
+            raise ImportError("To support decoding 'mp3' audio files, please install 'torchaudio'.") from err
+        if version.parse(torchaudio.__version__) < version.parse("0.12.0"):
+            try:
+                torchaudio.set_audio_backend("sox_io")
+            except RuntimeError as err:
+                raise ImportError("To support decoding 'mp3' audio files, please install 'sox'.") from err
+            array, sampling_rate = self._decode_mp3_torchaudio(path_or_file)
+        else:
+            try:  # try torchaudio anyway because sometimes it works (depending on the os and os packages installed)
+                array, sampling_rate = self._decode_mp3_torchaudio(path_or_file)
+            except RuntimeError:
+                try:
+                    # flake8: noqa
+                    import librosa
+                except ImportError as err:
+                    raise ImportError(
+                        "Your version of `torchaudio` (>=0.12.0) doesn't support decoding 'mp3' files on your machine. "
+                        "To support 'mp3' decoding with `torchaudio>=0.12.0`, please install `ffmpeg>=4` "
+                        'or downgrade `torchaudio` to <0.12: `pip install "torchaudio<0.12"`. '
+                        "To support decoding 'mp3' audio files without `torchaudio`, please install `librosa`: "
+                        "`pip install librosa`. Note that decoding will be extremely slow in that case."
+                    ) from err
+                # try to decode with librosa for torchaudio>=0.12.0 as a workaround
+                logger.warning("Decoding mp3 with `librosa` instead of `torchaudio`, decoding is slow.")
```

Comment:
Maybe we should warn this only once ? You can use `warnings.warn` to do that

Reply:
I can't figure out how to do this. I changed `logger.warning` to `warnings.warn` here and tried setting `warnings.filterwarnings("once")` / `warnings.simplefilter("once")` but it didn't work, What am I missing?

Reply:
@

Reply:
@lhoestq

Reply:
ah lol I used warnings from logger to check if decoding was done with librosa, and now warnings from .warn are not captured in `caplog` in pytest, so these tests fail

Reply:
I think it does the warning always once by default, you don't need to use an extra filter.

And I think `with pytest.warns(UserWarning):` should work in the test

Reply:
thanks! fixed it here https://github.com/huggingface/datasets/pull/4923/commits/ebf77d742db1e72386ae4b3d57c843d64cbefc6d

I don't know why but for some reason in librosa case warnings are shown at each decoding anyway, I checked it on Colab, see pic. Might it be because of that `librosa.load` itself gives warnings on decoding: `UserWarning: PySoundFile failed. Trying audioread instead.` 

Anyway, maybe not that important?

![image](https://user-images.githubusercontent.com/16348744/191261867-cb6882ba-d53d-40cc-9b97-6d5f211af531.png)

Reply:
Users can still mute those warnings, I don't think we can do something about it.